core/src/apps/tezos/sign_tx.py

```diff
@@ -215,3 +253,77 @@ def _encode_ballot(w: bytearray, ballot):
     write_uint32_be(w, ballot.period)
     write_bytes(w, ballot.proposal)
     write_uint8(w, ballot.ballot)
+
+
+def _encode_natural(w: bytearray, num):
+    # encode a natural integer with its signed bit on position 7
+    # as we do not expect negative numbers in a transfer operation the bit is never set
+    natural_tag = 0
+    write_uint8(w, natural_tag)
+
+    byte = num & 63
+    modified = num >> 6
+
+    if modified == 0:
+        write_uint8(w, byte)
+    else:
+        write_uint8(w, 128 | byte)
+        _encode_zarith(w, modified)
+
+
+def _encode_kt_common(w: bytearray, sequence_length, operation):
+    ADDRESS_LENGTH = 21
+
+    argument_length = (
+        sequence_length + 5
+    )  # 5 = tag and sequence_length (1 byte + 4 bytes)
+
+    helpers.write_bool(w, True)
+    write_uint8(w, helpers.DO_ENTRYPOINT_TAG)
+    write_uint32_be(w, argument_length)
+    write_uint8(w, helpers.MICHELSON_SEQUENCE_TAG)
+    write_uint32_be(w, sequence_length)
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["DROP"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["NIL"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["operation"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES[operation]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["key_hash"]))
+    if operation == "PUSH":
+        write_bytes(w, bytes([10]))  # byte sequence
+        write_uint32_be(w, ADDRESS_LENGTH)
+
+
+def _encode_smart_contract_transfer(w: bytearray, smart_contract_transfer):
+    MICHELSON_LENGTH = 48
+
+    value_natural = bytearray()
+    _encode_natural(value_natural, smart_contract_transfer.amount)
+    sequence_length = MICHELSON_LENGTH + len(value_natural)
+
+    _encode_kt_common(w, sequence_length, "PUSH")
+    write_bytes(w, smart_contract_transfer.recipient)
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["IMPLICIT_ACCOUNT"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["PUSH"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["mutez"]))
+    _encode_natural(w, smart_contract_transfer.amount)
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["UNIT"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["TRANSFER_TOKENS"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["CONS"]))
+
+
+# smart_contract_delegation
+def _encode_smart_contract_delegation(w: bytearray, smart_contract_delegation):
+    MICHELSON_LENGTH = 42  # length is fixed this time(no variable length fields)
+
+    _encode_kt_common(w, MICHELSON_LENGTH, "PUSH")
+    write_bytes(w, smart_contract_delegation.delegate)
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["SOME"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["SET_DELEGATE"]))
+    write_bytes(w, bytes(helpers.MICHELSON_INSTRUCTION_BYTES["CONS"]))
+
+
+def _encode_smart_contract_delegation_remove(w: bytearray, smart_contract_delegation):
```

Comment:
this function does not use the `smart_contract_delegation` argument. it should be removed